examples/rpc.exs

```diff
@@ -0,0 +1,165 @@
+      {:ok, spawn_link(fn -> init(procedure, params, options) end)}
+    end
+
+    # Private Functions
+
+    defp init(procedure, params, options) do
+      {:ok, caller} = RPC.new_peer([Spell.Role.Caller], options)
+      %__MODULE__{caller: caller,
+                  procedure: procedure,
+                  params: params}
+        |> struct(options)
+        |> loop()
+    end
+
+    defp loop(state) do
+
+
+      # `Spell.cast_call` performs an asyncronous call to the remote procedure, the result will 
+      # be intercepted and parsed by the block `Spell.receive_result`
+      {:ok, call_id} = Spell.cast_call(state.caller, state.procedure, state.params)
+      Logger.info("<Caller> send params: #{inspect(state.params)}")
+      
+      case Spell.receive_result(state.caller, call_id) do
+        {:ok, result} -> IO.inspect handle_result(result)
+        {:error, reason} -> {:error, reason}
+      end
+
+      :timer.sleep(1000)
+      loop(state)
+    end
+
+    def handle_result([_result_id, _options, _arguments, %{"result" => res}]) do
+      Logger.info("<Caller> received result: #{inspect(res)}")
+    end
+  end
+
+  defmodule Callee do
+    @moduledoc """
+    The callee need to be registered to the WAMP Dealer in order to be accesseble and expose
+    the procedure. Once it is registered it will wait an invocation from the Dealer and yield 
+    the result back. in this case it will receive an array of integer `[1, 2, 3]` and return
+    the sum of them `6`
+
+    iex> Callee.start_link("com.spell.rpc.sum")
+    """
+    require Logger
+    defstruct [:callee, :procedure, timeout: 1000]
+
+    # Public Functions
+
+    @doc """
+    Start the callee with the passed procedure.
+    """
+    def start_link(topics, options \\ []) do
+      {:ok, spawn_link(fn -> init(topics, options) end)}
+    end
+
+    # Private Functions
+
+    defp init(procedure, options) do
+      {:ok, callee} = RPC.new_peer([Spell.Role.Callee], options)
+
+      # `Spell.cast_register` receive the callee and the procedure where it
+      # has to be registered
+      {:ok, _register_id} = Spell.cast_register(callee, procedure)
+
+      %__MODULE__{callee: callee,
+                  procedure: procedure}
+        |> struct(options)
+        |> loop()
+    end
+
+    defp loop(state) do
+      receive do
+        :stop -> :ok
+        {Spell.Peer, _pid, message} -> handle_message(state, message)
+        _ -> {:error, :wrong_message}
+      after
+        state.timeout -> {:error, :timeout}
+      end
+      loop(state)
+    end
+
+    defp handle_message(_state, %Spell.Message{args: _args, code: _code, type: :registered}) do
+      Logger.info("<Callee> registered")
+    end
+
+    defp handle_message(state, %Spell.Message{args: [request, _reg_id, _msg, params], code: _code, type: :invocation}) do
+      Logger.info("<Callee> received #{inspect(params)} to sum")
+      Spell.cast_yield(state.callee, request, [arguments_kw: %{result: sum(params, 0)}])
```

Comment:
Instead of hard wiring the procedure to sum, you can pattern match on the procedure and call the regarding function dynamically. It maybe more obvious if you add another call with a different procedure.